File: src/main/java/com/futureclient/module/Module.java
package com.futureclient.module;

import com.futureclient.FutureClient;
import com.futureclient.config.Setting;
import net.minecraft.client.MinecraftClient;
import org.lwjgl.glfw.GLFW;
import java.util.ArrayList;
import java.util.List;

/**
 * Base class for all client modules
 */
public abstract class Module {
    protected final MinecraftClient mc = MinecraftClient.getInstance();
    private final String name;
    private final String description;
    private final Category category;
    private final List<Setting> settings = new ArrayList<>();
    
    private boolean enabled = false;
    private int keyBind = GLFW.GLFW_KEY_UNKNOWN;
    
    public Module(String name, String description, Category category) {
        this.name = name;
        this.description = description;
        this.category = category;
    }
    
    public void toggle() {
        setEnabled(!enabled);
    }
    
    public void setEnabled(boolean enabled) {
        this.enabled = enabled;
        if (enabled) {
            onEnable();
        } else {
            onDisable();
        }
        FutureClient.getInstance().getConfigManager().saveConfig();
    }
    
    public void onEnable() {}
    public void onDisable() {}
    public void onTick() {}
    public void onRender2D(float tickDelta) {}
    public void onRender3D(float tickDelta) {}
    
    protected void addSetting(Setting setting) {
        settings.add(setting);
    }
    
    public String getName() { return name; }
    public String getDescription() { return description; }
    public Category getCategory() { return category; }
    public boolean isEnabled() { return enabled; }
    public int getKeyBind() { return keyBind; }
    public void setKeyBind(int key) { this.keyBind = key; }
    public List<Setting> getSettings() { return settings; }
    
    public enum Category {
        COMBAT("Combat"),
        VISUAL("Visual"),
        MOVEMENT("Movement"),
        PLAYER("Player"),
        RENDER("Render"),
        HUD("HUD");
        
        private final String name;
        
        Category(String name) {
            this.name = name;
        }
        
        public String getName() { return name; }
    }
}

File: src/main/kotlin/com/futureclient/api/module/Module.kt
package com.futureclient.api.module

import net.minecraft.client.Minecraft
import net.minecraftforge.common.MinecraftForge

enum class Category {
    COMBAT, RENDER, PLAYER, MISC, HUD
}

abstract class Setting<T>(val name: String, var value: T)
class BooleanSetting(name: String, value: Boolean) : Setting<Boolean>(name, value)
class NumberSetting(name: String, value: Double, val min: Double, val max: Double, val increment: Double) : Setting<Double>(name, value)

abstract class Module(val name: String, val category: Category) {
    protected val mc: Minecraft = Minecraft.getMinecraft()
    var toggled: Boolean = false
    var key: Int = 0
    val settings = ArrayList<Setting<*>>()

    fun toggle() {
        toggled = !toggled
        if (toggled) onEnable() else onDisable()
    }

    open fun onEnable() { MinecraftForge.EVENT_BUS.register(this) }
    open fun onDisable() { MinecraftForge.EVENT_BUS.unregister(this) }
    protected fun addSetting(setting: Setting<*>) { settings.add(setting) }
}
